frontend: Settings: Replace theme dropdown menu with grid-based selector

AllanSlomski opened the pull request:
This PR replaces the theme <Select> dropdown with a grid-based selector.

The grid uses appThemes to render all themes visually, allowing users to click and apply them.

I noticed the base themes are defined in themeSlice.ts, but I couldn’t clearly identify how to distinguish them from plugin-registered themes. If there’s a recommended way to separate base and plugin themes visually, I’d love some guidance to follow up with a second PR or improvement.

![headlamp](https://github.com/user-attachments/assets/07cd771a-e259-4676-b074-c84aa9c91857)

#3155

sniok commented on frontend/src/components/App/Settings/Settings.tsx:
```diff
@@ -72,24 +72,36 @@ export default function Settings() {
-                      {capitalize(it.name)}
-                    </Box>
-                  </MenuItem>
+                  <Box
```
Using Box (which renders to a div) for an interactive element is not accessible, a better option for this element would be a Button

AllanSlomski replied:
I initially tried using Button, but its default properties were overriding the custom styles and configurations I needed. Because of that, I opted to use Box with the role="button" instead. Do you think this is a reasonable approach, or would you recommend another way to handle this?

sniok replied:
yeah this is okay

sniok commented on frontend/src/components/App/Settings/Settings.tsx:
```diff
@@ -72,24 +72,36 @@ export default function Settings() {
+                    key={it.name}
+                    sx={{
+                      cursor: 'pointer',
+                      border: themeName === it.name ? '2px solid' : '1px solid #ccc',
```
using hardcoded colors (#ccc) is not recommended, for a border color you could use `theme.palette.divider `

sniok commented on frontend/src/components/App/Settings/Settings.tsx:
```diff
@@ -72,24 +72,36 @@ export default function Settings() {
+                      flexDirection: 'column',
+                      alignItems: 'center',
+                      transition: '0.2 ease',
+                      '&:hover': {
```
In headlamp we don't really use shadows for hover status, to be more consistent with the rest of the app I'd suggest instead changing background and/or border color

sniok commented on frontend/src/components/App/Settings/Settings.tsx:
```diff
@@ -72,24 +72,36 @@ export default function Settings() {
+                    }}
+                    onClick={() => dispatch(setTheme(it.name))}
+                  >
+                    <ThemePreview theme={it} />
```
ThemePreview should be a bit bigger, you can use the size prop

sniok reviewed (changes requested):
Thanks for opening a PR! This is a good start! 
We had in mind to have a theme selector to be on the bottom of the settings page and take full width, something similar to a github theme selector

![image](https://github.com/user-attachments/assets/65b3cdb5-3b4d-4776-8389-b8c30bc4e34e)

I've also left a couple of comments regarding some small issues

Appreciate you taking a look into it

AllanSlomski commented:
Thank you! I'm working in these changes now. I'll update the PR as soon as possible.

AllanSlomski commented:
I made all the changes you requested:

I used role="button" on the Box tag because using the Button component directly would override my styles and configurations.

I used the divider color for the border.

I removed the boxShadow and added a background color using divider as well.

I added a prop to change the ThemePreview's size.

The core change is the design: I moved the Theme Selector to the bottom of the page, as you suggested. To make it look better, I placed it outside the table — otherwise, I wouldn't be able to properly customize all the properties I wanted, especially to make the subtitle and theme cards look similar to GitHub's appearance page.

If there's anything else that needs adjustment, please let me know. Thank you!

sniok commented on frontend/src/components/App/Settings/Settings.tsx:
```diff
@@ -127,6 +104,83 @@ export default function Settings() {
+                }}
+                sx={{
+                  cursor: 'pointer',
+                  border: themeName === it.name ? '3px solid' : '1px solid',
```
for "active" state let's use a primary color (theme.palette.primary) for the 1px border so it provides more contrast. In the current version it is a bit hard to see which item is selected.

sniok reviewed (changes requested):
Nice, thanks for the updates! I think this is very close to being ready. One thing that I think would improve the look a bit is if the items are aligned to the left instead of being stretched.

Also a snapshot test is failing but I think there's a bug that makes it produce different test results based on the OS. If you reverse changes to pluginLib.snapshot I think the tests will pass in CI.

AllanSlomski commented:
Thanks for the feedback! I'm already working on the changes you suggested, but I'm not sure if I understood the snapshot error. The snapshots are updated automatically when I run the tests. Should I replace it with the old one, from before my changes, and dont update before commiting?

sniok commented:
> Thanks for the feedback! I'm already working on the changes you suggested, but I'm not sure if I understood the snapshot error. The snapshots are updated automatically when I run the tests. Should I replace it with the old one, from before my changes, and dont update before commiting?

Since there's an issue with pluginLib.snapshot generation you should exclude changes to that file from the PR (replace it with the old one). All the other snapshots updates are okay

AllanSlomski commented:
Hello!
As you suggested, I changed the border colors and aligned the items to the left.
I made different breakpoints for smaller screens.
For the snapshot error, I copied the file from the original repository and replaced it, so it's now using the original version.
Additionally, there are some issues that I've noticed, and I'm putting screenshots of them here. 

![Captura de Tela (135)](https://github.com/user-attachments/assets/f237474c-1093-4869-938d-4371d09cf95a)
![Captura de Tela (137)](https://github.com/user-attachments/assets/62b874dd-7548-4311-ad1b-419a9052cf49)

sniok commented:
Looks great! @AllanSlomski thank you for the updates!
Just a couple of small things left, could you rebase your branch and make sure that license header is still present in frontend/src/components/App/Settings/Settings.tsx?
There's an empty package-lock in the root directory that got accidentally committed
And could you also squash all the changes into a single commit
There are still some changes in the pluginLib.snapshot, tests are not passing

sniok commented:
> About the snapshot errors: I tried copying the original file back into my code, but the tests are still failing. Could you please clarify what I should do to resolve this issue properly?

You could use `git checkout` to get the snapshot file from main branch before creating a commit. 
Since this feature is something that we want to include in the release pretty soon, would you mind if I took over this PR since there are just a couple of small changes left?

AllanSlomski commented:
Thanks!
I'll be commiting with the changes as soon as I can.

AllanSlomski commented:
Do you mind waiting for a few hours or just until tomorrow?

AllanSlomski commented:
Hello!

I believe I accidentally squashed a commit that wasn't mine — the one that updates the translation files. I'm really sorry about that.

Should I try to fix it, or would you prefer to handle it directly?

Thanks again for your help and feedback!

illume commented:
Thanks @AllanSlomski 🎉
